aggregated-dependency-score.go

```diff
@@ -60,21 +60,39 @@ type DependencyResolver interface {
 	GetDirectDependencies(ctx context.Context, p Package) ([]Package, error)
 }
 
-func (eval *trustwhorthinessEvaluator) evaluate(ctx context.Context, p Package) (float64, error) {
-	intrinsic, err := eval.intrinsic.EvaluateIntrinsicTrustworthiness(ctx, p)
+func (evaluator *trustwhorthinessEvaluator) evaluate(ctx context.Context, p Package, ancestors map[string]struct{}) (float64, error) {
```

Comment:
I think you are tackling the "[directed acyclic graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph)" problem here. Would be interesting to see whether golang provides an idiomatic solution for topological sorting and then just walk the sorted result

Reply:
ChatGPT on resolving this:
```golang
package main

import "fmt"

// ListUniqueNodesFromRoot traverses the graph from a single root node and lists all nodes uniquely.
func ListUniqueNodesFromRoot(graph map[string][]string, root string) []string {
	visited := make(map[string]bool)
	var uniqueNodes []string

	// Helper function for DFS
	var dfs func(node string)
	dfs = func(node string) {
		if visited[node] {
			return
		}
		visited[node] = true
		uniqueNodes = append(uniqueNodes, node)
		for _, neighbor := range graph[node] {
			dfs(neighbor)
		}
	}

	// Start DFS from the root
	dfs(root)

	return uniqueNodes
}

func main() {
	// Example graph as an adjacency list
	graph := map[string][]string{
		"PackageA": {"PackageB", "PackageC"},
		"PackageB": {"PackageD"},
		"PackageC": {"PackageD", "PackageE"},
		"PackageD": {"PackageF"},
		"PackageE": {},
		"PackageF": {},
	}

	// Specify the root package
	root := "PackageA"

	// Get all unique nodes reachable from the root
	uniqueNodes := ListUniqueNodesFromRoot(graph, root)
	fmt.Println("Unique Nodes:", uniqueNodes)
}
```

Reply:
I first thought that you are avoiding to run the same dependency twice but you are just trying to avoid loops which makes me wonder whether you want to account for the same package multiple times if it occurs multiple times as a dependency

Reply:
that would seem to overweight standard deps. Say a moderate vuln in the stdlib package fmt, that would be re-accounted for each dependency